rust/windows-client/src-tauri/src/client/gui.rs

```diff
@@ -5,22 +5,51 @@
 
 use crate::client::{self, deep_link, AppLocalDataDir};
 use anyhow::{anyhow, bail, Context, Result};
+use arc_swap::ArcSwap;
 use client::settings::{self, AdvancedSettings};
-use connlib_client_shared::file_logger;
+use connlib_client_shared::{file_logger, ResourceDescription};
 use secrecy::{ExposeSecret, SecretString};
 use std::{
     net::{Ipv4Addr, Ipv6Addr},
     path::PathBuf,
     str::FromStr,
+    sync::Arc,
 };
 use tauri::{
     CustomMenuItem, Manager, SystemTray, SystemTrayEvent, SystemTrayMenu, SystemTrayMenuItem,
     SystemTraySubmenu,
 };
-use tokio::sync::{mpsc, oneshot};
+use tokio::sync::{mpsc, oneshot, Notify};
 use ControllerRequest as Req;
 
-pub(crate) type CtlrTx = mpsc::Sender<ControllerRequest>;
+/// CtlrTx allows the controller task to only await a single `Notify` even though it
+/// can receive messages from unrelated sources like connlib's `on_update_resources`
+/// and the `mpsc` channel that gets system tray events.
+#[derive(Clone)]
+pub(crate) struct CtlrTx {
+    ctlr_tx: mpsc::Sender<ControllerRequest>,
+    // The controller only waits for its Notify, so use that to wake it up
+    // Tokio docs say that Notify has the same inter-thread ordering as stdlib's
+    // thread parking, so it's safe to use Notify to wake up a task to read unrelated data.
+    // Tokio will synchronize and make sure that data written before `notify_*` is readable before the other task wakes.
+    notify_controller: Arc<Notify>,
```

Comment:
If you are using a `mpsc` channel the `Notify` seems redundant.

Reply:
It's so I can wait on the ArcSwap

Reply:
I thought I had documented that here but maybe it was in another comment

Reply:
ah, so it's the same `Notify`? you could instead do

```rs
tokio::select! {
   notify_controller.notified() => { // handle the notify }
   ctrl = ctrl_tx.recv() => { // handle the queue }
}
```
(I think you suggested that in another comment already, I just think we should implement that in this PR it's a bit confusing otherwise)

Reply:
Yeah I think I'll do that. Something about `select!` being a macro that's limited to 64 options made me wonder if the implementation inside it is strange, but it'll work, and then I can delete `struct CtlrTx`

Reply:
Did that